Fix #1043 - make `CompileNodeToValue` filesystem-agnostic

Ocramius opened the pull request:
With this change, `CompileNodeToValue` will honor the source location given by the source locator, and will not use the local filesystem to identify `__DIR__` nor `__FILE__`.

Note: as it currently is, this patch is just a stub of the final wished result, as some cleanup is still necessary.

Also, this should remove some filesystem calls that could happen deep inside `roave/better-reflection`'s core, potentially speeding up the whole thing :D 

Fixes #1043

Ocramius committed:
Fix #1043 - make `CompileNodeToValue` filesystem-agnostic

With this change, `CompileNodeToValue` will honor the source location given
by the source locator, and will not use the local filesystem to identify
`__DIR__` nor `__FILE__`.

Note: as it currently is, this patch is just a stub
of the final wished result, as some cleanup is still
necessary.

Fixes #1043

Ocramius commented on test/unit/NodeCompiler/CompileNodeToValueTest.php:
```diff
@@ -678,12 +683,108 @@ public function magicConstantsInNamespaceProvider(): array
 
+    public function testCanRetrieveMagicConstantValueWhenUsingFakeSourceLocator(): void
+    {
+        $astLocatorWithFakeFilePath = new class ($this->astLocator) extends Locator {
```
Not a fan of this declaration - potentially moving it to a new file?

Ocramius replied:
Decided to keep it here, and fixed variable name.

Ocramius commented on test/unit/NodeCompiler/CompileNodeToValueTest.php:
```diff
@@ -678,12 +683,108 @@ public function magicConstantsInNamespaceProvider(): array
     public function testMagicConstantsInNamespace(string $constantName, mixed $expectedValue): void
     {
-        $reflector = new DefaultReflector(new SingleFileSourceLocator(__DIR__ . '/../Fixture/MagicConstants.php', $this->astLocator));
+        $reflector = new DefaultReflector(new SingleFileSourceLocator(realpath(__DIR__ . '/../Fixture/MagicConstants.php'), $this->astLocator));
```
Minor breakage here, but the file path should still be valid (even if relative), so IMO this is an OK change for 5.3.0

asgrim reviewed (approved):
Seems decent :+1:

Ocramius commented on test/unit/NodeCompiler/CompileNodeToValueTest.php:
```diff
@@ -678,12 +683,108 @@ public function magicConstantsInNamespaceProvider(): array
+        );
+
+        self::assertSame(
+            '/non/existing/path/to/sources.php',
```
Not sure whether this test will "just work" on windows.

Ocramius replied:
Confirmed by @kukulich that this "just works"

kukulich reviewed (approved):
Tests are green on Windows so 👍

Ocramius committed:
Minor cleanups - removed unused inputs, removed excessive test code, removed excessive stubs

This should clarify the intent of the anonymous class implementing `Locator` in a test too.

Ocramius commented:
Now in `5.3.0` :)
